### DISCUSSION TITLE
JS evaluate methods

### DISCUSSION THREAD

**Discussion by Yarflam** [pyscript/pyscript#1614] [Q&A] [answered]
Hello,

Is it possible to run code as if in a virtual environment? Such like `pyscript.vm.runInNewContext("print('x:', x)", { x: 3 })` or just like `pyscript.eval("print('ok')")`.

Thank you!

edit: if it's not possible, I would know how to change the source code (how it's working today?).

**Comment by WebReflection** [comment, marked as answer]
Once you have the PyScript interpreter you can `interpreter.runPython(...)` and obtain results, somehow demoed in this discussion: https://github.com/pyscript/pyscript/discussions/1587#discussioncomment-6424731

To have a new context I'm afraid right now PyScript classic doesn't offer that, but PyScript next will ... or better, its core, *polyscript*, already does that via the `env` attribute: https://pyscript.github.io/polyscript/

**Comment by Yarflam** [comment]
All right, thank you very much for your answer!

I think it's too complicated to use this library with a JS framework. Pyodide looks like more accessible ([test codepen](https://codepen.io/Yarflam/pen/yLQQQvN)), not perfect (no local scope for variables, no verbose mode etc) but I can interact with it directly.

**Reply by WebReflection** [reply]
wow ... "*it's too complicated*" is exactly the kind of project scope we're trying to avoid/solve so I am sad you think that but if you look closer, this is all it takes to have your example up and running: https://codepen.io/WebReflection/pen/oNQJbpJ?editors=1000

```html
<script type="module" src="https://esm.sh/polyscript"></script>
<script type="pyodide" config='{"packages":["numpy"]}'>
import numpy as np

d = np.int64(15.34)

def test(x):
    return x*2

print("x:", test(d))

def test(x):
    return x*3

print("x:", test(d))
</script>
```

I haven't written a single line of JS and the console outputs the same ... I hope *polyscript* and *PyScript* will be looked as "*this is too simple*" instead!

**Reply by Yarflam** [reply]
Okay, yes but if you want you to execute this code from a text area? User change the code and click on "Execute" button, is it possible? Thanks.

Because for my project behind, I need to implement in React Bundle: https://codepen.io/Yarflam/pen/MWzGBXP
Today, the user can execute a JS code inside ; it's the main concept, you don't need setup Jupyter on your computer and configure it for Javascript. If I can execute Python inside, it's a bonus (I can't create app better than Google Colab and I don't want) - it's not open source project, but I give free access online and reusable (component can be import on own website).

**Reply by WebReflection** [reply]
https://codepen.io/WebReflection/pen/eYQbZvQ?editors=1000

```html
<script type="module">
import { env } from 'https://esm.sh/polyscript';

const interpreter = await env.pyodide;

// you can execute this via textarea content too
interpreter.runPython(`
import numpy as np

d = np.int64(15.34)

def test(x):
    return x*2

print("x:", test(d))

def test(x):
    return x*3

print("x:", test(d))
`);
</script>

<!-- still need to define a desired interpreter + config -->
<script type="pyodide" config='{"packages":["numpy"]}'></script>
```

**Reply by Yarflam** [reply]
Fine, I made it with Pyodide: https://codepen.io/Yarflam/pen/vYQbBvK

Pyodide environment is loaded in iframe. When the user clicks on the button "Restart Kernel", the iframe is closed and recreated.
Then simply capture the returns, console.log/warn/error (overrides) and distribute them.

I've the same performance problem as PyScript/Polyscript: loading Pyodide blocks the JS execution of the page. I will look to replace iframe by service worker (https://pyodide.org/en/stable/usage/webworker.html#web-worker).

**Reply by WebReflection** [reply]
You can't have `js.document` in Service Worker, with *polyscript* you can via `XWorker` ... the example would then be:

```html
<script type="module">
import { XWorker } from 'https://esm.sh/polyscript';

new XWorker('./your-project.py', {
  type: 'pyodide',
  config: {
    packages: ['numpy']
  }
});
</script>
```

In `your-project.py` you then have:

```python
# [ if needed ]
# from polysript import xworker
# document = xworker.window.document

import numpy as np

d = np.int64(15.34)

def test(x):
    return x*2

print("x:", test(d))

def test(x):
    return x*3

print("x:", test(d))
```

This loads *pyodide* **only** in the worker and it won't ever block the main thread. You have `xworker.window` which proxies all things `pyodide.js` would but through the worker ... I hope this example unlocks your project.

**edit** P.S. to use Atomics and SharedArrayBuffer you need either [mini-coi](https://github.com/WebReflection/mini-coi#readme) as Service Worker or you need to provide the right headers via your server ... I use `npx static-handler --cors --coop --coep --corp .` to ensure those headers are there in *localhost* but for live static pages you might consider the *mini-coi* solution.
